README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->



# MCHT

*Version 0.2.0.9000*

[![GitHub issues](https://img.shields.io/github/issues/ntguardian/MCHT.svg)](https://github.com/ntguardian/MCHT/issues)
[![GitHub forks](https://img.shields.io/github/forks/ntguardian/MCHT.svg)](https://github.com/ntguardian/MCHT/network)
[![GitHub stars](https://img.shields.io/github/stars/ntguardian/MCHT.svg)](https://github.com/ntguardian/MCHT/stargazers)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![Github All Releases](https://img.shields.io/github/downloads/ntguardian/MCHT/total.svg)](https://github.com/ntguardian/MCHT)

**MCHT** is a package implementing an interface for creating and using Monte
Carlo tests. The primary function of the package is `MCHTest()`, which creates
functions with S3 class `MCHTest` that perform a Monte Carlo test.

## Installation

**MCHT** is not presently available on CRAN. You can download and install 
**MCHT** from GitHub using [**devtools**](https://github.com/r-lib/devtools) via
the R command `devtools::install_github("ntguardian/MCHT")`.

## Monte Carlo Hypothesis Testing

Monte Carlo testing is a form of hypothesis testing where the <img src="svgs/2ec6e630f199f589a2402fdf3e0289d5.svg" align=middle width=8.270567249999992pt height=14.15524440000002pt/>-values are
computed using the empirical distribution of the test statistic computed from
data simulated under the null hypothesis. These tests are used when the
distribution of the test statistic under the null hypothesis is intractable or
difficult to compute, or as an exact test (that is, a test where the
distribution used to compute <img src="svgs/2ec6e630f199f589a2402fdf3e0289d5.svg" align=middle width=8.270567249999992pt height=14.15524440000002pt/>-values is appropriate for any sample size, not
just large sample sizes).

Suppose that <img src="svgs/aabe1517ce1102595512b736cbf264bb.svg" align=middle width=15.831502499999988pt height=14.15524440000002pt/> is the observed value of the test statistic and large values
of <img src="svgs/aabe1517ce1102595512b736cbf264bb.svg" align=middle width=15.831502499999988pt height=14.15524440000002pt/> are evidence against the null hypothesis; normally, <img src="svgs/2ec6e630f199f589a2402fdf3e0289d5.svg" align=middle width=8.270567249999992pt height=14.15524440000002pt/>-values would be
computed as <img src="svgs/94e4cf2543ecdf13ca181360434546e6.svg" align=middle width=70.60317329999998pt height=24.65753399999998pt/>, where <img src="svgs/1ef429296c00d0d4dc9914bfb2f6ec6f.svg" align=middle width=147.44866949999997pt height=24.65753399999998pt/> is the cumulative
distribution functions and <img src="svgs/49aebd2501b0bf3a5225ca26ba123672.svg" align=middle width=18.205948199999987pt height=22.465723500000017pt/> is the random variable version of <img src="svgs/aabe1517ce1102595512b736cbf264bb.svg" align=middle width=15.831502499999988pt height=14.15524440000002pt/>. We
cannot use <img src="svgs/b8bc815b5e9d5177af01fd4d3d3c2f10.svg" align=middle width=12.85392569999999pt height=22.465723500000017pt/> for some reason; it's intractable, or the <img src="svgs/b8bc815b5e9d5177af01fd4d3d3c2f10.svg" align=middle width=12.85392569999999pt height=22.465723500000017pt/> provided is only
appropriate for large sample sizes.

Instead of using <img src="svgs/b8bc815b5e9d5177af01fd4d3d3c2f10.svg" align=middle width=12.85392569999999pt height=22.465723500000017pt/> we will use <img src="svgs/15c3c9c70eb47be5a6e886765530f5d7.svg" align=middle width=22.21695959999999pt height=31.141535699999984pt/>, which is the empirical CDF of
the same test statistic computed from simulated data following the distribution
prescribed by the null hypothesis of the test. For the sake of simplicity in
this presentation, assume that <img src="svgs/e257acd1ccbe7fcb654708f1a866bfe9.svg" align=middle width=11.027402099999989pt height=22.465723500000017pt/> is a continuous random variable. Now our
<img src="svgs/2ec6e630f199f589a2402fdf3e0289d5.svg" align=middle width=8.270567249999992pt height=14.15524440000002pt/>-value is <img src="svgs/422c7ea56f597c35a6c675a532225eb9.svg" align=middle width=80.78808375pt height=31.141535699999984pt/>, where <img src="svgs/fa268ea8502a266665526053011ee08d.svg" align=middle width=206.580495pt height=32.19743999999999pt/> where <img src="svgs/21fd4e8eecd6bdf1a4d3d6bd1fb8d733.svg" align=middle width=8.484300000000001pt height=22.381919999999983pt/> is the indicator function and
<img src="svgs/bc3c694d37b92361e3102381d7c007e6.svg" align=middle width=28.21459079999999pt height=30.267491100000004pt/> is an independent random copy of <img src="svgs/49aebd2501b0bf3a5225ca26ba123672.svg" align=middle width=18.205948199999987pt height=22.465723500000017pt/> computed from simulated
data with a sample size of <img src="svgs/55a049b8f161ae7cfeb0197d75aff967.svg" align=middle width=9.86687624999999pt height=14.15524440000002pt/>.

The power of these tests increase with <img src="svgs/f9c4988898e7f532b9f826a75014ed3c.svg" align=middle width=14.99998994999999pt height=22.465723500000017pt/> (see [1]) but modern computers are
able to simulate large <img src="svgs/f9c4988898e7f532b9f826a75014ed3c.svg" align=middle width=14.99998994999999pt height=22.465723500000017pt/> quickly, so this is rarely an issue. The procedure
above also assumes that there are no nuisance parameters and the distribution of
<img src="svgs/49aebd2501b0bf3a5225ca26ba123672.svg" align=middle width=18.205948199999987pt height=22.465723500000017pt/> can effectively be known precisely when the null hypothesis is true (and
all other conditions of the test are met, such as distributional assumptions). A
different procedure needs to be applied when nuisance parameters are not
explicitly stated under the null hypothesis. [2] suggests a procedure using
optimization techniques (recommending simulated annealing specifically) to
adversarially select values for nuisance parameters valid under the null
hypothesis that maximize the <img src="svgs/2ec6e630f199f589a2402fdf3e0289d5.svg" align=middle width=8.270567249999992pt height=14.15524440000002pt/>-value computed from the simulated data. This
procedure is often called *maximized Monte Carlo* (MMC) testing. That is the
procedure employed here. (In fact, the tests created by `MCHTest()` are the
tests described in [2].) Unfortunately, MMC, while conservative and exact, has
much less power than if the unknown parameters were known, perhaps due to the
behavior of samples under distributions with parameter values distant from the
true parameter values (see [3]).

## Bootstrap Hypothesis Testing

Bootstrap statistical testing is very similar to Monte Carlo testing; the key
difference is that bootstrap testing uses information from the sample. For
example a parametric bootstrap test would estimate the parameters of the
distribution the data is assumed to follow and generate datasets from that
distribution using those estimates as the actual parameter values. A permutation
test (like Fisher's permutation test; see [4]) would use the original dataset
values but randomly shuffle the labeles (stating which sample an observation
belongs to) to generate new data sets and thus new simulated test statistics.
<img src="svgs/2ec6e630f199f589a2402fdf3e0289d5.svg" align=middle width=8.270567249999992pt height=14.15524440000002pt/>-values are essentially computed the same way.

Unlike Monte Carlo tests and MMC, these tests are not exact tests. That said,
they often have good finite sample properties. (See [3].)

See the documentation for more details and references.

## Examples

`MCHTest()` is the main function of the package and can create functions with S3
class `MCHTest` that perform Monte Carlo hypothesis tests.

For example, the code below creates the Monte Carlo equivalent of a <img src="svgs/4f4f4e395762a3af4575de74c019ebb5.svg" align=middle width=5.936097749999991pt height=20.221802699999984pt/>-test.


```r
library(MCHT)
#> .------..------..------..------.
#> |M.--. ||C.--. ||H.--. ||T.--. |
#> | (\/) || :/\: || :/\: || :/\: |
#> | :\/: || :\/: || (__) || (__) |
#> | '--'M|| '--'C|| '--'H|| '--'T|
#> `------'`------'`------'`------' v. 0.1.0
#> Type citation("MCHT") for citing this R package in publications
library(doParallel)
#> Loading required package: foreach
#> Loading required package: iterators
#> Loading required package: parallel

registerDoParallel(detectCores())  # Necessary for parallelization, and if not
                                   # done the resulting function will complain
                                   # on the first use

ts <- function(x, mu = 0) {sqrt(length(x)) * (mean(x) - mu)/sd(x)}
sg <- function(x, mu = 0) {
  x <- x + mu
  ts(x)
}
rg <- rnorm

mc.t.test <- MCHTest(ts, sg, rg, seed = 20181001, test_params = "mu", 
                     lock_alternative = FALSE,
                     method = "Monte Carlo One Sample t-Test")
```

The object `mc.t.test()` is an S3 class, and a callable function.


```r
class(mc.t.test)
#> [1] "MCHTest"
```

`print()` will print relevant information about the construction of the test.


```r
mc.t.test
#> 
#> 	Details for Monte Carlo One Sample t-Test
#> 
#> Seed:  20181001 
#> Replications:  10000 
#> Tested Parameters:  mu 
#> Default mu:  0 
#> 
#> Memoisation enabled
```

Once this object is created, we can use it for performing hypothesis tests.


```r
dat <- c(2.3, -0.13, 1.42, 1.51, 3.43, -0.96, 0.59, 0.62, 1.28, 4.07)

t.test(dat, mu = 1, alternative = "two.sided")  # For reference
#> 
#> 	One Sample t-test
#> 
#> data:  dat
#> t = 0.84975, df = 9, p-value = 0.4175
#> alternative hypothesis: true mean is not equal to 1
#> 95 percent confidence interval:
#>  0.3135303 2.5124697
#> sample estimates:
#> mean of x 
#>     1.413
mc.t.test(dat, mu = 1)
#> Loading required package: rngtools
#> Loading required package: pkgmaker
#> Loading required package: registry
#> 
#> Attaching package: 'pkgmaker'
#> The following object is masked from 'package:base':
#> 
#>     isFALSE
#> 
#> 	Monte Carlo One Sample t-Test
#> 
#> data:  dat
#> S = 0.84975, p-value = 0.9885
mc.t.test(dat, mu = 1, alternative = "two.sided")
#> 
#> 	Monte Carlo One Sample t-Test
#> 
#> data:  dat
#> S = 0.84975, p-value = 0.023
#> alternative hypothesis: true mu is not equal to 1
```

This is the simplest example; `MCHTest()` can create more involved Monte Carlo
tests. See other documentation for details.

## Planned Future Features

* A function for making diagnostic-type plots for tests, such as a function
    creating a plot for the rejection probability function (RPF) as described in
    [5]
* A function that accepts a `MCHTest`-class object and returns a function that,
    rather than returning a `htest`-class object, returns a function that will
    give the test statistic, simulated test statistics, and a <img
    src="svgs/2ec6e630f199f589a2402fdf3e0289d5.svg" align=middle
    width=8.270567249999992pt height=14.15524440000002pt/>-value, in a list;
    could be useful for diagnostic work.

## References

1. A. C. A. Hope, *A simplified Monte Carlo test procedure*, JRSSB, vol. 30
   (1968) pp. 582-598
2. J-M Dufour, *Monte Carlo tests with nuisance parameters: A general approach
   to finite-sample inference and nonstandard asymptotics*, Journal of
   Econometrics, vol. 133 no. 2 (2006) pp. 443-477
3. J. G. MacKinnon, *Bootstrap hypothesis testing* in *Handbook of computational
   econometrics* (2009) pp. 183-213
4. R. A. Fisher, *The design of experiments* (1935)
5. R. Davidson and J. G. MacKinnon, *The size distortion of bootstrap test*,
   Econometric Theory, vol. 15 (1999) pp. 361-376


NEWS.md
# MCHT News

## Version 0.2.0
---

### Functions

- Created `stat_t()`, the test statistic for the t-test
- Created `stat_t_two_sample()`, the test statistic for Welch's two-sample
    t-test

## Version 0.1.0
---

- Created `MCHTest()`, which creates `MCHTest`-class objects that handle a
    particular Monte Carlo or bootstrap test
- Created `pval()`, a function for computing p-values for tests based on
    empirical distributions, and supports breaking ties via generating uniformly
    distributed random variables
- Create a print method for `MCHTest`-class objects that prints information
    about the parameters of the object
- Created `is.MCHTest()` function to check whether an object inherits from the
    `MCHTest` class
- Created `get_MCHTest_settings()` which creates a list with all objects that
    together uniquely characterize an `MCHTest`-class object

## Version 0.0.0.9000
---

### Setup

- Package **MCHT** created.


tests/testthat.R
library(testthat)
library(MCHT)

test_check("MCHT")


tests/testthat/test-MCHTestChecks.R
################################################################################
# test-MCHTestChecks.R
################################################################################
# 2018-09-29
# Curtis Miller
################################################################################
# Tests for MCHTest-related functions
################################################################################

library(MCHT)
library(doParallel)
context("MCHTest Checks")

dat <- c(0.09, 1.36, 1.44, 0.71, 0.02, 0.34, 0.14, 0.44, 0.6, 0.16)

test_stat_1 <- function(x) {
  sqrt(length(x)) * mean(x)/sd(x)
}

test_stat_2 <- function(x, mu) {
  sqrt(length(x)) * (mean(x) - mu)/sd(x)
}

test_stat_3 <- function(x, mu, sigma) {
  sqrt(length(x)) * (mean(x) - mu)/sigma
}

test_stat_4 <- function(x, mu) {
  mean(x) - mu
}

test_gen_1 <- function(x) {
  sqrt(length(x)) * mean(x) / sd(x)
}

test_gen_2 <- function(x, mu) {
  x <- x + mu
  sqrt(length(x)) * (mean(x) - mu)/sd(x)
}

test_gen_3 <- function(x, mu, sigma) {
  x <- sigma * x + mu
  sqrt(length(x)) * (mean(x) - mu)/sd(x)
}

test_gen_4 <- function(x, mu, sigma) {
  x <- sigma * x + mu
  mean(x) - mu
}

rand_gen_1 <- function(n) {
  rnorm(n)
}

test_that("MCHTest() functions properly", {
  expect_is(MCHTest(test_stat_1, test_stat_1, rand_gen_1, seed = 100),
            "MCHTest")
  expect_true(is.MCHTest(MCHTest(test_stat_1, test_stat_1, rand_gen_1,
                                 seed = 100)))
  expect_warning(MCHTest(test_stat_1, test_stat_1, rand_gen_1))
  expect_error(MCHTest(test_stat_1, test_stat_1, rand_gen_1, seed = 100,
                       test_params = "mu"))
  expect_error(MCHTest(test_stat_1, test_gen_2, rand_gen_1, seed = 100,
                       test_params = "mu"))
  expect_error(MCHTest(test_stat_2, test_gen_2, rand_gen_1,
                       test_params = c("mu", "sigma"), seed = 100))
  expect_silent(MCHTest(test_stat_2, test_gen_2, rand_gen_1, test_params = "mu",
                        seed = 10))
  expect_silent(MCHTest(test_stat_1, test_stat_1, rand_gen_1,
                        memoise_sample = FALSE))
  expect_error(MCHTest(test_stat_2, test_gen_1, rand_gen_1, seed = 100,
                       test_params = "mu", fixed_params = "sigma"))
  expect_silent(MCHTest(test_stat_3, test_gen_3, rand_gen_1, seed = 100,
                        test_params = "mu", fixed_params = "sigma"))
  expect_error(MCHTest(test_stat_4, test_gen_3, rand_gen_1, seed = 100,
                       test_params = "mu", nuisance_params = "sigma"))
  expect_silent(MCHTest(test_stat_4, test_gen_3, rand_gen_1, seed = 100,
                        test_params = "mu", nuisance_params = "sigma",
                        optim_control = list(lower = c("sigma" = 0),
                                             upper = c("sigma" = 100))))
  expect_error(MCHTest(test_stat_4, test_gen_3, rand_gen_1, seed = 100,
                       test_params = "mu", nuisance_params = "sigma",
                       optim_control = list(lower = c(0),
                                            upper = c(100))))
  expect_error(MCHTest(test_stat_3, test_gen_3, rand_gen_1, seed = 100,
                       test_params = "mu", fixed_params = "sigma",
                       nuisance_params = "sigma", optim_control = list(
                         lower = c("sigma" = 0), upper = c("sigma" = 100)
                       )), "Parameter vectors cannot")
  expect_error(MCHTest(test_stat_1, rnorm, rand_gen_1, seed = 100))
  expect_silent(MCHTest(test_stat_1, test_gen_1, function(x) {x}, seed = 100))
  expect_error(MCHTest(test_stat_1, test_gen_1, function(y) {y}, seed = 100))
})

test_that("An MCHTest-class object functions properly", {
  mc.test.1 <- MCHTest(test_stat_1, test_stat_1, rand_gen_1, seed = 100,
                       N = 1000)
  mc.test.2 <- MCHTest(test_stat_1, test_stat_1, rand_gen_1, N = 1000,
                       memoise_sample = FALSE)
  mc.test.3 <- MCHTest(test_stat_2, test_gen_2, rand_gen_1, N = 100, seed = 100,
                       test_params = "mu", lock_alternative = FALSE)
  mc.test.4 <- MCHTest(test_stat_3, test_gen_3, rand_gen_1, N = 100, seed = 100,
                       test_params = "mu", fixed_params = "sigma",
                       lock_alternative = FALSE)
  mc.test.5 <- MCHTest(test_stat_4, test_gen_4, rand_gen_1, N = 100, seed = 100,
                       test_params = "mu", nuisance_params = "sigma",
                       optim_control = list("lower" = c("sigma" = 0),
                                            "upper" = c("sigma" = 100)),
                       lock_alternative = FALSE, threshold_pval = 0.2)
  mc.test.6 <- MCHTest(test_stat_4, test_gen_4, rand_gen_1, N = 100, seed = 100,
                       test_params = "mu", nuisance_params = "sigma",
                       optim_control = list("lower" = c("sigma" = 0),
                                            "upper" = c("sigma" = 100)),
                       lock_alternative = FALSE, threshold_pval = 0.2,
                       suppress_threshold_warning = TRUE)

  set.seed(1234)
  registerDoParallel(1)
  expect_success(mc.test.1(dat))
  expect_error(mc.test.3(dat), "argument \"mu\" is missing")
  expect_true(mc.test.2(dat - 0.5)$p.value != mc.test.2(dat - 0.5)$p.value)
  expect_equal(mc.test.1(dat)$statistic, c(S = 3.28607799050981))
  expect_equal(mc.test.1(dat)$p.value, 0.005)
  expect_is(mc.test.1(dat), "htest")
  expect_equal(mc.test.1(dat)$data.name, "dat")
  expect_equal(mc.test.1(dat), mc.test.1(dat, alternative = "two.sided"))
  expect_equal(mc.test.3(dat, alternative = "two.sided", mu = 0.5)$p.value,
               0.94)
  expect_equal(mc.test.3(dat, alternative = "greater", mu = 0.5)$p.value,
               0.47)
  expect_equal(mc.test.3(dat, alternative = "greater", mu = 0.5)$alternative,
               "greater")
  expect_error(MCHTest(test_stat_1, test_stat_1, rand_gen_1, seed = 100, N = 5,
                       pval_func = function(S, sample_S) {2})(dat),
               "Bad p-value")
  expect_equal(mc.test.4(dat, alternative = "two.sided", mu = 0.5,
                         sigma = 1)$p.value, 0.96)
  expect_equal(mc.test.4(dat, alternative = "two.sided", mu = 0.5,
                         sigma = 2)$p.value, 0.94)
  expect_warning(mc.test.5(dat, alternative = "two.sided", mu = 0.5),
                 "Computed p-value is greater than threshold value \\(0.2\\);")
  expect_silent(mc.test.6(dat, alternative = "two.sided", mu = 0.5))
  expect_equal(mc.test.6(dat, alternative = "two.sided", mu = 0.5)$p.value,
               0.9)

  ts <- function(x, sigma = 1) {sqrt(length(x)) * mean(x)/sigma}
  sg <- function(x, sigma = 1) {x <- sigma * x; ts(x, sigma = sigma)}
  rg <- function(n) {rnorm(n)}
  mc.test.7 <- MCHTest(ts, sg, rg, seed = 123, N = 100, fixed_params = "sigma")
  mc.test.8 <- MCHTest(ts, sg, rg, seed = 123, N = 100, fixed_params = "sigma",
                       localize_functions = TRUE,
                       imported_objects = list("ts" = ts))
  ts <- function(x) {mean(x)}

  expect_error(mc.test.7(dat, sigma = 2))
  expect_success(mc.test.8(dat, sigma = 2))

  mc.test.9 <- MCHTest(test_stat_1, function(x) {test_stat(x)},
                       function(x) {rnorm(length(x))}, seed = 123, N = 100,
                       localize_functions = TRUE)
  expect_success(mc.test.9(dat))
})

test_that("get_MCHTest_settings() functions properly", {
  f_setting <- get_MCHTest_settings(MCHTest(test_stat_2, test_gen_2, rand_gen_1,
                                            seed = 100, memoise_sample = TRUE,
                                            N = 10000, test_params = "mu",
                                            lock_alternative = FALSE))

  expect_equal(f_setting$seed, 100)
  expect_equal(f_setting$N, 10000)
  expect_true(f_setting$memoise_sample)
  expect_equal(f_setting$method, "Monte Carlo Test")
  expect_equal(f_setting$test_params, "mu")
  expect_false(f_setting$lock_alternative)
  expect_is(f_setting$test_stat, "function")
  expect_is(f_setting$stat_gen, "function")
  expect_is(f_setting$rand_gen, "function")
  expect_is(f_setting$pval_func, "function")
})


tests/testthat/test-StatisticsChecks.R
################################################################################
# test-StatisticsChecks.R
################################################################################
# 2018-09-30
# Curtis Miller
################################################################################
# Check that statistics functions work properly
################################################################################

library(MCHT)
context("Check statistical functions")

dat <- c(0.86, 2.62, 0.59, 0.84, 1.55, 4.06, 1.83, 0.67, 0.4, 0.92)
df1 <- data.frame("v1" = c(dat, dat + 2, 1),
                  "v2" = rep(c("x", "y", "z"),
                             times = c(length(dat), length(dat), 1)))

test_that("pval works properly", {
  set.seed(123)
  expect_equal(pval(5, 1:10, unif_gen = runif), 0.4)
  expect_equal(pval(5, 1:10, alternative = "greater", unif_gen = runif), 0.5)
  expect_equal(pval(6, 1:10, alternative = "less", unif_gen = runif), 0.7)
  expect_equal(pval(6, 1:10, alternative = "two.sided", unif_gen = runif), 0.6)
  expect_error(pval(6, 1:10, alternative = "none", unif_gen = runif))
  expect_equal(pval(5, c(1:10, 5, 5, 5),
                    unif_gen = MCHT:::gen_memo_rng(runif)),
               0.230769230769231)
  expect_equal(pval(5, c(1:10, 5, 5, 5),
                    unif_gen = MCHT:::gen_memo_rng(runif, 123)),
               0.307692307692308)
  expect_equal(pval(5, c(1:10, 5, 5, 5),
                    unif_gen = MCHT:::gen_memo_rng(runif, 123)),
               0.307692307692308)
  expect_equal(pval(5, c(1:10, 5, 5, 5),
                    unif_gen = MCHT:::gen_memo_rng(runif)),
               0.153846153846154)
  expect_equal(pval(6, 1:10), 0.4)
})

test_that("Test statistic functions work properly", {
  expect_equal(stat_t(dat, mu = 2), -1.56416733137808)
  expect_error(stat_t_two_sample(df1, id_var = "v2",
                                 value_var = "v1"),
               "Not two unique groups in v2")
  expect_equal(stat_t_two_sample(df1[1:20,], delta = -3, id_var = "v2",
                                 value_var = "v1"), sqrt(5)/sd(dat))
})


R/Utils.R
################################################################################
# Utils.R
################################################################################
# 2018-08-27
# Curtis Miller
################################################################################
# Functions and other extras used throughout the package
################################################################################

################################################################################
# ROXYGEN2 TAGS
################################################################################

################################################################################
# OPERATORS
################################################################################

#' Concatenate (With Space)
#'
#' Concatenate and form strings (with space separation)
#'
#' @param x One object
#' @param y Another object
#' @return A string combining \code{x} and \code{y} with a space separating them
#' @examples
#' `%s%` <- MCHT:::`%s%`
#' "Hello" %s% "world"
`%s%` <- function(x, y) {paste(x, y)}

#' Concatenate (Without Space)
#'
#' Concatenate and form strings (no space separation)
#'
#' @inheritParams %s%
#' @return A string combining \code{x} and \code{y}
#' @examples
#' `%s0%` <- MCHT:::`%s0%`
#' "Hello" %s0% "world"
`%s0%` <- function(x, y) {paste0(x, y)}

################################################################################
# PARAMETER CHECKING
################################################################################

#' Check That Parameters Are In Functions
#'
#' Test that certain parameters are arguments for certain functions via
#' \pkg{testthat} functions.
#'
#' @param params Character vector with parameter names
#' @param func_list List of functions to check
#' @examples
#' MCHT:::check_params_in_functions(c("x"), list(mean))
check_params_in_functions <- function(params, func_list) {
  testthat::expect_is(params, "character")
  testthat::expect_is(func_list, "list")
  for (f in func_list) {
    testthat::expect_is(f, "function")
    testthat::expect_true(all(params %in% names(formals(f))) |
      ("..." %in% names(formals(f))))
  }
}


R/StatisticalFunctions.R
################################################################################
# StatisticalFunctions.R
################################################################################
# 2018-09-29
# Curtis Miller
################################################################################
# Functions for statistical tasks.
################################################################################

################################################################################
# P-VALUE COMPUTATION
################################################################################

#' Compute \eqn{p}-Value For a Test Statistic
#'
#' Compute the \eqn{p}-value of a test statistic for Monte Carlo tests.
#'
#' Let \eqn{S} be a test statistic and \eqn{S_i} be simulated values of that
#' test statistic under the null hypothesis, with \eqn{1 \leq i \leq N}. If
#' \code{unif_gen} is not \code{NULL}, this function computes \eqn{p}-values via
#'
#' \deqn{p = \hat{p} = \frac{1}{N}\sum_{i = 1}^{N} I_{\{(S, U_0) \leq
#' (S_i, U_i)\}}}
#'
#' where \eqn{I_{\{S \in A\}} = 1} if \eqn{S \in A} and is 0
#' otherwise, \eqn{U_i} are uniformly distributed random variables, and the
#' ordering over tuples is lexicographical ordering, as described by
#' \insertCite{dufour06;textual}{MCHT}.
#'
#' If \code{unif_gen} is \code{NULL}, then the random variables are not
#' generated and not used to break ties.
#'
#' This function is designed to handle an \code{alternative} parameter similar
#' to what appears in other \pkg{stats} functions like
#' \code{\link[stats]{t.test}}. If \code{alternative} is \code{"less"}, then
#' \eqn{p = \hat{p}}; if \code{alternative} is \code{"greater"}, then
#' \eqn{p = 1 - \hat{p}}; and if \code{alternative} is \code{"two.sided"}, then
#' \eqn{p = 2 \min(\hat{p}, 1 - \hat{p})}. Any other value raises an error.
#'
#' The parameter \code{S} is \eqn{S}, and the vector \code{sample_S} is the
#' vector containing the values \eqn{S_i}.
#'
#' @param S The value of the test statistic
#' @param sample_S Simulated values of the 
#' @param alternative A string specifying the alternative hypothesis, or
#'                    \code{NULL}
#' @param unif_gen If not \code{NULL}, the function generating
#'                 uniformly-distributed random variables for breaking ties; if
#'                 \code{NULL}, no tie breaking is done
#' @return A number representing the \eqn{p}-value.
#' @references
#'   \insertAllCited{}
#' @export
#' @examples
#' sample_S <- rnorm(10)
#' pval(1.01, sample_S)
#' pval(1.01, sample_S, alternative = "greater")
pval <- function(S, sample_S, alternative = NULL, unif_gen = NULL) {
  if (!is.null(unif_gen)) {
    check_params_in_functions("n", list(unif_gen))
  }

  N <- length(sample_S)
  sample_S <- as.numeric(sample_S)
  if (is.null(alternative)) {
    alternative <- "greater"
  }

  empirical_cdf_value <- mean(sample_S <= S)
  if (is.function(unif_gen)) {
    tiebreakers <- unif_gen(N + 1)
    empirical_cdf_value <- empirical_cdf_value + mean(
      (tiebreakers[1:N] <= tiebreakers[N + 1]) & (S == sample_S))
  }
  if (alternative == "greater") {
    1 - empirical_cdf_value
  } else if (alternative == "less") {
    empirical_cdf_value
  } else if (alternative == "two.sided") {
    2 * min(empirical_cdf_value, 1 - empirical_cdf_value)
  } else {
    stop("Don't know how to interpret alternative")
  }
}

################################################################################
# RANDOM VARIABLE GENERATION
################################################################################

#' Memoised Random Variable Generation
#'
#' Creates a function that generates random numbers with memoization
#'
#' This is a function generator, with the returned function being one that can
#' handle a set seed and will remember if it needs to regenerate a new set of
#' random numbers. This allows both for control over random number generation
#' and for faster performance.
#'
#' @param r The random number generator
#' @param seed The seed value, to be passed to \code{\link[base]{set.seed}}
#' @return A function that generates random numbers with a set seed and with
#'         memoization; accepts \code{seed} and all other arguments that could
#'         be passed to the original random number generator
#' @examples
#' memo_runif <- MCHT:::gen_memo_rng(runif)
#' memo_runif(10)
gen_memo_rng <- function(r, seed = NULL) {
  force(r)
  force(seed)
  old_seed <- NULL
  out <- NULL
  old_args <- NULL
  
  f <- function(...) {
    args <- list(...)
    if (!exists(".Random.seed")) {
      stats::runif(1)  # Call a random number that won't be used; initiates RNG
    }
    if (!is.null(seed)) {
      old_seed <<- .Random.seed
      set.seed(seed)
    }
    # Check for change in random state or whether function's been called or
    # whether call itself has changed
    if (is.null(out) | any(.Random.seed != old_seed) |
        !identical(args, old_args)) {
      # Regenerate random data
      out <<- do.call(r, args)
      old_args <<- args
      # Enables memoization even if seed not set
      if (is.null(seed) | is.null(old_seed)) {
        old_seed <<- .Random.seed
      }
    }
    if (!is.null(seed)) {
      .Random.seed <<- old_seed
    }
    out
  }
}

################################################################################
# USEFUL TEST STATISTICS
################################################################################

#' One-Sample t Statistic
#'
#' The one-sample \eqn{t}-statistic.
#'
#' Let \eqn{\mu} correspond to the parameter \code{mu} representing the mean
#' under the null hypothesis. Then the \eqn{t}-statistic is
#'
#' \deqn{\frac{\bar{x} - \mu}{s/\sqrt{n}}}
#'
#' where \eqn{\bar{x}} is the sample mean, \eqn{s} is the sample standard
#' deviation, and \eqn{n} is the sample size. \insertNoCite{student08}{MCHT}
#'
#' @param x The dataset for which to compute the \eqn{t} statistic
#' @param mu The mean under the null hypothesis
#' @return Number representing the value of the \eqn{t}-statistic
#' @references
#'   \insertAllCited{}
#' @export
#' @examples
#' stat_t(rnorm(10), mu = 1)
stat_t <- function(x, mu = 0) {
  sqrt(length(x)) * (mean(x) - mu)/sd(x)
}

#' Two-Sample t Statistic
#'
#' Test statistic for Welch's two-sample \eqn{t}-test
#' \insertCite{welch47}{MCHT}.
#'
#' Let \eqn{\Delta} correspond to the difference in the population means of the
#' two samples under the null hypothesis. Welch's two-sample \eqn{t} statistc is
#'
#' \deqn{\frac{\bar{x} - \bar{y} - \Delta}{\sqrt{\frac{s_x^2}{n_x} +
#' \frac{s_y^2}{n_y}}}}
#'
#' where \eqn{\bar{x}} and \eqn{\bar{y}} are the means of the two respective
#' samples, \eqn{s_x} and \eqn{s_y} are the standard devations of the two
#' respective samples, and \eqn{n_x} and \eqn{n_y} are the sizes of the two
#' samples.
#'
#' The function parameter \code{x} does not correspond to the sample \eqn{x} as
#' described above but to a data frame that contains both samples, with a
#' variable representing the value of the data points of the samples and a
#' variable identifying the sample to which a sample belongs.
#'
#' Which sample is considered the \eqn{x} sample and which the \eqn{y} sample in
#' the above calculation is determined by lexicographical order.
#'
#' @param x A data frame with two columns, \code{id_var} and \code{value_var},
#'          and no others
#' @param delta The difference in population means under the null hypothesis
#' @param id_var The name of the variable identifying which sample an
#'               observation (row of the data frame) belongs to; the column must
#'               either contain character data or a \code{\link[base]{factor}}
#'               with two levels identifying the sample (will be coerced to
#'               \code{factor} if character)
#' @param value_var The numeric variable representing observation values
#' @return Number representing the value of the statistic
#' @references
#'   \insertAllCited{}
#' @examples
#' df <- data.frame("value" = c(rnorm(10), rnorm(5) + 1),
#'                  "group" = rep(c("x", "y"), times = c(10, 5)))
#' stat_t_two_sample(df)
stat_t_two_sample <- function(x, delta = 0, id_var = "group",
                              value_var = "value") {
  testthat::expect_identical(sort(names(x)), sort(c(id_var, value_var)))
  x[[id_var]] <- as.character(x[[id_var]])
  x[[id_var]] <- as.factor(x[[id_var]])
  if (length(levels(x[[id_var]])) != 2) {stop("Not two unique groups in" %s%
                                               id_var)}
  x <- x[, c(value_var, id_var)]
  names(x) <- c("value", "group")

  means <- aggregate(value ~ group, data = x, FUN = mean)
  sigma2s <- aggregate(value ~ group, data = x, FUN = var)
  sizes <- aggregate(value ~ group, data = x, FUN = length)

  means <- means[order(as.character(means$group)), "value"]
  sigma2s <- sigma2s[order(as.character(sigma2s$group)), "value"]
  sizes <- sizes[order(as.character(sizes$group)), "value"]

  (means[1] - means[2] - delta)/sqrt(sum(sigma2s/sizes))
}

# TODO: curtis: CHI-SQUARE STATISTIC INDEPENDENCE -- Mon 15 Oct 2018 01:17:27 PM MDT
# TODO: curtis: CHI-SQUARE STATISTIC VARIANCE -- Mon 15 Oct 2018 01:17:27 PM MDT
# TODO: curtis: F-STATISTIC TWO SAMPLE VARIANCE -- Mon 15 Oct 2018 01:17:27 PM MDT
# TODO: curtis: F-STATISTIC LINEAR REGRESSION -- Mon 15 Oct 2018 01:17:27 PM MDT
# TODO: curtis: CHI-SQUARE STATISTIC GOODNESS-OF-FIT -- Mon 15 Oct 2018 01:17:27 PM MDT
# TODO: curtis: ANOVA STATISTIC -- Mon 15 Oct 2018 01:17:27 PM MDT
